File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
private:
	bool check(vector<int>& bloomDay, int day, int m, int k) {
		int ct = 0, ans = 0;
		for (int i = 0; i < bloomDay.size(); ++i) {
            
			if (bloomDay[i] <= day) ct++;
			else ct = 0;
            
			if (ct == k) {
				ans++;
				ct = 0;
			}
		}
		return ans >= m;
	}
public:
	int minDays(vector<int>& bloomDay, int m, int k) {
		int lo = 1, hi = 1e9 + 1, ans = -1;

		while (lo <= hi) {
			int mid = (lo + hi) >> 1;

			if (check(bloomDay, mid, m, k)) {
				ans = mid;
				hi = mid - 1;
			}
			else lo = mid + 1;
		}
		return ans;
	}
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int start = 1;
        int end = Arrays.stream(bloomDay).max().getAsInt();
        int ans = -1;
        
        while (start <= end) {
            int mid = start + (end - start) / 2;
            int currLength = 0;
            int count = 0;
            
            for (int i = 0; i < bloomDay.length; i++) {
                if (bloomDay[i] <= mid) {
                    currLength++;
                    if (currLength >= k) {
                        currLength = 0;
                        count++;
                    }
                } else {
                    currLength = 0;
                }
            }
            
            if (count >= m) {
                ans = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        return ans;
    }
}